File: 279/279_version1.py
class Solution:
    def numSquares(self, n: int) -> int:
        if n <= 1:
            return n
        result = [0, 1]
        number = 2
        while number <= n:
            value = float("inf")
            option = int(number ** 0.5)
            while option > 0 and result[number-option**2]+1 <= value:
                value = min([value, result[number - option * option] + 1])
                option -= 1
            if value != float("inf"): result.append(value)
            number += 1
        return result[-1]


print(Solution().numSquares(48))

File: 279/279_version1.java
class Solution {
    public int numSquares(int n) {
        if(n<=1) return n;
        List<Integer> result = new ArrayList<>();
        result.add(0);
        result.add(1);
        int number = 2;
        while(number <= n){
            int value = number;
            for(int reduceNum = 1;number-reduceNum*reduceNum>=0;reduceNum++){
                value = Math.min(value,result.get(number-reduceNum*reduceNum)+1);
            }
            result.add(value);
            number+=1;
        }
        return result.get(n);
    }
}
